// @adobe/lit-mobx/demo/detaching-example.js
/*
Copyright 2018 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { property, customElement } from 'lit/decorators.js';
import { cache } from 'lit/directives/cache.js';
import { MobxLitElement } from '../src/lit-mobx';
import { Counter } from './my-counter';
import './public-property-example.js';
let DetachingExample = class DetachingExample extends MobxLitElement {
    constructor() {
        super(...arguments);
        this.counter = new Counter();
        this.connectChild = false;
    }
    render() {
        return html `
            <h3>Connect/Disconnect Example</h3>
            <p>
                In this example we allow an element to be connected/disconnected
                and show that it is correctly updated after re-connection.
            </p>
            ${cache(this.connectChild
            ? html `
                          <public-observable-example
                              .counter=${this.counter}
                          ></public-observable-example>
                      `
            : nothing)}
            <br />
            <button @click=${this.incrementCount}>Add</button>
            <button @click=${this.toggleConnected}>Connect/Disconnect</button>
        `;
    }
    incrementCount() {
        this.counter.increment();
    }
    toggleConnected() {
        this.connectChild = !this.connectChild;
    }
};
__decorate([
    property({ type: Boolean })
], DetachingExample.prototype, "connectChild", void 0);
DetachingExample = __decorate([
    customElement('detaching-example')
], DetachingExample);
export { DetachingExample };
